doc/build-helpers/images/dockertools.section.md

```diff
@@ -902,102 +902,192 @@ The `name` argument is the name of the derivation output, which defaults to `fro
+
+Since it is a script, this should be used with other dockerTools functions in attributes that expect scripts.
+See [](#ex-dockerTools-shadowSetup-buildImage) and [](#ex-dockerTools-shadowSetup-buildLayeredImage) for how to do that.
+
+:::{.note}
+It is recommended to use either `fakeNss` or `shadowSetup` depending on your use case, but not both.
+:::
+
+### Examples {#ssec-pkgs-dockerTools-helpers-examples}
+
+:::{.example #ex-dockerTools-helpers-buildImage}
+# Using `dockerTools`'s environment helpers with `buildImage`
+
+This example adds the [`binSh`](#sssec-pkgs-dockerTools-helpers-binSh) helper to a basic Docker image built with [`dockerTools.buildImage`](#ssec-pkgs-dockerTools-buildImage).
+Without this helper, it wouldn't be possible to enter a shell inside a container running this image.
+This is the `buildImage` equivalent of [](#ex-dockerTools-helpers-buildLayeredImage).
 
 ```nix
-buildImage {
-  name = "shadow-basic";
+{ dockerTools, hello }:
+dockerTools.buildImage {
+  name = "env-helpers";
+  tag = "latest";
 
-  runAsRoot = ''
-    #!${pkgs.runtimeShell}
-    ${pkgs.dockerTools.shadowSetup}
-    groupadd -r redis
-    useradd -r -g redis redis
-    mkdir /data
-    chown redis:redis /data
-  '';
-}
+  copyToRoot = [
+    hello
+    dockerTools.binSh
+  ];
 ```
 
-Creating base files like `/etc/passwd` or `/etc/login.defs` is necessary for shadow-utils to manipulate users and groups.
+After building the image and loading it in Docker, we can create a container based on it and enter a shell inside the container.
+This wouldn't be possible without `binSh`.
+
+```shell
+$ nix-build
+(some output removed for clarity)
+/nix/store/2p0i3i04cgjlk71hsn7ll4kxaxxiv4qg-docker-image-env-helpers.tar.gz
+$ docker load -i /nix/store/2p0i3i04cgjlk71hsn7ll4kxaxxiv4qg-docker-image-env-helpers.tar.gz
+(output removed for clarity)
+$ docker run --rm -it env-helpers:latest /bin/sh
+sh-5.2# help
+GNU bash, version 5.2.21(1)-release (x86_64-pc-linux-gnu)
+(rest of output removed for clarity)
+```
+:::
+
+:::{.example #ex-dockerTools-helpers-buildLayeredImage}
+# Using `dockerTools`'s environment helpers with `buildLayeredImage`
+
+This example adds the [`binSh`](#sssec-pkgs-dockerTools-helpers-binSh) helper to a basic Docker image built with [`dockerTools.buildLayeredImage`](#ssec-pkgs-dockerTools-buildLayeredImage).
+Without this helper, it wouldn't be possible to enter a shell inside a container running this image.
+This is the `buildLayeredImage` equivalent of [](#ex-dockerTools-helpers-buildImage).
 
-When using `buildLayeredImage`, you can put this in `fakeRootCommands` if you `enableFakechroot`:
 ```nix
-buildLayeredImage {
-  name = "shadow-layered";
+{ dockerTools, hello }:
+dockerTools.buildLayeredImage {
+  name = "env-helpers";
+  tag = "latest";
 
-  fakeRootCommands = ''
-    ${pkgs.dockerTools.shadowSetup}
-  '';
-  enableFakechroot = true;
+  contents = [
```

Comment:
Not sure that we should recommend this. `contents` has too much implicit behavior for my taste at least.
By making the choice of merging logic explicit, we inform the reader (and our readers' code's readers) of the implementation choice, which they may well have to change.
`symlinkJoin` and `buildEnv` are quite different.

Reply:
Can you elaborate on this? I'm a bit confused, because there are a couple things going on in my mind:
- `buildImage` uses `copyToRoot` and has `contents` as deprecated, but this example uses `buildLayeredImage` which only provides `contents` for putting stuff into the image. Did you make this comment with this in mind already?
- If you did, I wasn't aware of any scenarios where `contents` in `buildLayeredImage` would cause issues due to implicit behaviour. Can you give me some pointers to learn more about this? This seems like something we definitely want to have documented in the manual, but currently this is not mentioned at all.

Reply:
It seems fine to me, let's merge this for now and improve the section with future PRs :)